# Mobile App For Lottery Addiction

A medical institute that aims to prevent and treat gambling addictions wants to build a dedicated mobile app to help lottery addicts better estimate their chances of winning. 

The goal of this project is to create the logical core of the app and calculate probabilities.

## Core Functions

In [2]:
def factorial(n):
    final_product = 1
    for i in range(n, 0, -1):
        final_product *= i
    return final_product

def combinations(n, k):
    numerator = factorial(n)
    denominator = factorial(k) * factorial(n-k)
    return numerator/denominator
    

## One-ticket Probability

In [3]:
def one_ticket_probability(user_num):
    n_comb = combinations(49, 6)
    prob_one_ticket = 1/n_comb
    percent_one_ticket = prob_one_ticket * 100
    print ('''Your chances to win the jackpot with {} are {:.7f}%. 
    In other words you have a 1 in {:,} chance to win!'''.format(user_num,
    percent_one_ticket, int(n_comb)))

In [4]:
one_ticket_probability([1,2,30,4,5,6])

Your chances to win the jackpot with [1, 2, 30, 4, 5, 6] are 0.0000072%. 
    In other words you have a 1 in 13,983,816 chance to win!


The function above takes user inputs of 6 different numbers from 1 to 49, and return the probability of winning in a user friendly way.

## Historical Data Check For Canada Lottery

In [7]:
import pandas as pd

lotto = pd.read_csv("649.csv")
lotto.shape

(3665, 11)

In [8]:
lotto.head(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [9]:
lotto.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

The lotto 649 data set contains historical data for 3,665 drawings, dating from 1982 to 2018.

## Function for Historical Data Check

In [10]:
# Go over each row in lotto set and extract historical winning numbers 
def extract_numbers(row):
    row = row[4:10]
    row = set(row.values)
    return row

winning_num = lotto.apply(extract_numbers, axis=1)
winning_num.head(3)

0    {3, 41, 11, 12, 43, 14}
1    {33, 36, 37, 39, 8, 41}
2     {1, 6, 39, 23, 24, 27}
dtype: object

In [26]:
# Check to see if user set of numbers have ever occurred
def check_historical_occurrence(user_num, hist_num):
    user_num_set = set(user_num)
    check_occurrence = hist_num == user_num_set
    n_occurrences = check_occurrence.sum()
    if n_occurrences == 0:
        print('''The combination {} has never occurred. 
The chance to win with your combination {} at the next draw is 0.0000072%.
This is the same as saying you have a 1 in 13,983,816 chances to win.'''.format(user_num, user_num))
    else:
        print('''The number of times combination {} has occurred is {}.
The chance to win with your combination {} at the next draw is 0.0000072%.
This is the same as saying you have a 1 in 13,983,816 chances to win.'''.format(user_num, n_occurrences, user_num))
    

In [27]:
check_historical_occurrence([33, 36, 37, 39, 8, 41], winning_num)

The number of times combination [33, 36, 37, 39, 8, 41] has occurred is 1.
The chance to win with your combination [33, 36, 37, 39, 8, 41] at the next draw is 0.0000072%.
This is the same as saying you have a 1 in 13,983,816 chances to win.


In [28]:
check_historical_occurrence([3, 7, 11, 20, 24, 28], winning_num)

The combination [3, 7, 11, 20, 24, 28] has never occurred. 
The chance to win with your combination [3, 7, 11, 20, 24, 28] at the next draw is 0.0000072%.
This is the same as saying you have a 1 in 13,983,816 chances to win.


For the app the user inputs their 649 numbers. Under the hood, the 6 numbers serves as an input for the check_historical_occurrence function. The output shows how many times those numbers have won before and what the probability is of winning with those numbers in the next draw (it is always 1 in 13,983,816).

## Multi-ticket Probability

In [40]:
def multi_ticket_probability(n_tickets):
    n_comb = combinations(49, 6)
    probability = n_tickets/n_comb
    percentage = probability * 100 
    one_in_chance = round(n_comb/n_tickets)
    if n_tickets == 1:
        print('''Your chance to win the jackpot with one ticket is {:.6f}%.
This is the same as saying you have a 1 in {:,} chance to win.'''.format(percentage, int(n_comb)))
    else:
        print('''Your chance to win the jackpot with {:,} different numbers are {:.6f}%.
This is the same as saying you have a 1 in {:,} chance to win.'''.format(n_tickets, percentage, one_in_chance))

In [42]:
multi_ticket_probability(1)

Your chance to win the jackpot with one ticket is 0.000007%.
This is the same as saying you have a 1 in 13,983,816 chance to win.


In [43]:
multi_ticket_probability(10)

Your chance to win the jackpot with 10 different numbers are 0.000072%.
This is the same as saying you have a 1 in 1,398,382 chance to win.


In [44]:
multi_ticket_probability(100)

Your chance to win the jackpot with 100 different numbers are 0.000715%.
This is the same as saying you have a 1 in 139,838 chance to win.


In [45]:
multi_ticket_probability(10000)

Your chance to win the jackpot with 10,000 different numbers are 0.071511%.
This is the same as saying you have a 1 in 1,398 chance to win.


In [46]:
multi_ticket_probability(6991908)

Your chance to win the jackpot with 6,991,908 different numbers are 50.000000%.
This is the same as saying you have a 1 in 2 chance to win.


In [47]:
multi_ticket_probability(13983816)

Your chance to win the jackpot with 13,983,816 different numbers are 100.000000%.
This is the same as saying you have a 1 in 1 chance to win.


In the app the user should know what the probability of winning the jackpot is when they buy multiple tickets.

The user inputs how many numbers they intend to play. The function should print out numbers showing what the chances to win are.

## Non-Jackpot Winning Numbers 

In [49]:
def probability_less_6(n_winning_num):
    n_combinations_ticket = combinations(6, n_winning_num)
    n_combinations_remaining = combinations(49 - n_winning_num, 6 - n_winning_num)
    successful_outcomes = n_combinations_ticket * n_combinations_remaining
    n_combinations_total = combinations(49, 6)
    
    probability = successful_outcomes / n_combinations_total
    probability_percentage = probability * 100
    
    combinations_simplified = round(n_combinations_total/successful_outcomes)
    
    print('''Your chances of having {} winning numbers with this ticket are {:.6f}%.
In other words, you have a 1 in {:,} chances to win.'''.format(n_winning_num, probability_percentage,
                                                               int(combinations_simplified)))

In [50]:
for test_input in [2, 3, 4, 5]:
    probability_less_6(test_input)
    print('--------------------------') # output delimiter

Your chances of having 2 winning numbers with this ticket are 19.132653%.
In other words, you have a 1 in 5 chances to win.
--------------------------
Your chances of having 3 winning numbers with this ticket are 2.171081%.
In other words, you have a 1 in 46 chances to win.
--------------------------
Your chances of having 4 winning numbers with this ticket are 0.106194%.
In other words, you have a 1 in 942 chances to win.
--------------------------
Your chances of having 5 winning numbers with this ticket are 0.001888%.
In other words, you have a 1 in 52,969 chances to win.
--------------------------


In the 649 lotto, there are smaller prizes for hitting 2, 3, 4, or 5 of the 6 numbers. 
In the app, the user inputs 6 different numbers from 1 to 49, and an integer between 2 and 5 that represents the number of winning numbers expected.